### PR TITLE
(FACT-1934) Add ssh fact for Windows OpenSSH feature

### PR DESCRIPTION
**Pull request by GeorgeMrejea**

### CODE REVIEW

File: lib/src/facts/windows/ssh_resolver.cc [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        // Since there is no user associated with the sshd service, the host keys are stored under \ProgramData\ssh.
+        string sshPath = ((path(dataPath) / "ssh").string());
+        static vector<string> const search_directories = {
+                sshPath
+        };
+
+        // Search the directories for the fact's key file
+        path key_file;
+        for (auto const& directory : search_directories) {
+            key_file = directory;
```
Comment (GabrielNagy) [inline lib/src/facts/windows/ssh_resolver.cc]: I know you copied the code from the posix ssh_resolver, but we don't really need to iterate over a single directory here...

File: lib/src/facts/windows/ssh_resolver.cc [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+namespace facter { namespace facts { namespace windows {
+
+    ssh_resolver::data ssh_resolver::collect_data(collection& facts)  {
+            data result;
+            populate_key("ssh_host_rsa_key.pub", 1, result.rsa);
+            populate_key("ssh_host_dsa_key.pub", 2, result.dsa);
+            populate_key("ssh_host_ecdsa_key.pub", 3, result.ecdsa);
+            populate_key("ssh_host_ed25519_key.pub", 4, result.ed25519);
+            return result;
+        }
```
Comment (GabrielNagy) [inline lib/src/facts/windows/ssh_resolver.cc]: please use 4 spaces for indentation, that seems to be the standard here

File: lib/src/facts/windows/ssh_resolver.cc [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    void ssh_resolver::populate_key(std::string const& filename, int type, ssh_key& key)
+    {
+        string dataPath;
+        if (!environment::get("programdata", dataPath)) {
+                   LOG_DEBUG("error finding programdata: {1}",
+                           leatherman::windows::system_error());
+               }
```
Comment (GabrielNagy) [inline lib/src/facts/windows/ssh_resolver.cc]: please use 4 spaces for indentation, that seems to be the standard here

File: lib/src/facts/windows/ssh_resolver.cc (hunk 1 above)
Comment (GeorgeMrejea) [reply lib/src/facts/windows/ssh_resolver.cc]: Agree, for moment Windows allows for this feature only one directory, will remove iteration

File: lib/src/facts/windows/ssh_resolver.cc (hunk 3 above)
Comment (GeorgeMrejea) [reply lib/src/facts/windows/ssh_resolver.cc]: done

File: lib/src/facts/windows/ssh_resolver.cc (hunk 2 above)
Comment (GeorgeMrejea) [reply lib/src/facts/windows/ssh_resolver.cc]: done

File: lib/src/facts/windows/ssh_resolver.cc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        return result;
+    }
+
+    void ssh_resolver::populate_key(std::string const& filename, int type, ssh_key& key)
+    {
+        string dataPath;
+        if (!environment::get("programdata", dataPath)) {
+                LOG_DEBUG("error finding programdata: {1}",
+                leatherman::windows::system_error());
+        }
```
Comment (GabrielNagy) [inline lib/src/facts/windows/ssh_resolver.cc]: still have extra spaces here. also, you can fit the contents in one line:
`LOG_DEBUG("error finding programdata: {1}", leatherman::windows::system_error());`

File: lib/inc/internal/facts/windows/ssh_resolver.hpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#pragma once
```
Comment (ciprianbadescu) [inline lib/inc/internal/facts/windows/ssh_resolver.hpp]: This file is identical to the one in posix namespace. Can both me merge in a single file in parent directory?

File: lib/CMakeLists.txt [hunk 6]
```diff
@@ -117,6 +117,9 @@ if (WIN32)
         "src/facts/windows/collection.cc"
         "src/util/windows/wsa.cc"
     )
+    if (OPENSSL_FOUND)
```
Comment (ciprianbadescu) [inline lib/CMakeLists.txt]: this is now duplicate with condition at line 103, could this be executed once outside of UNIX/WIN32 checks?

File: lib/src/facts/windows/ssh_resolver.cc [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+using namespace boost::filesystem;
+
+namespace bs = boost::system;
+namespace lth_file = leatherman::file_util;
+
+using namespace leatherman::util;
+
+namespace facter { namespace facts { namespace windows {
+
+    ssh_resolver::data ssh_resolver::collect_data(collection& facts)
```
Comment (ciprianbadescu) [inline lib/src/facts/windows/ssh_resolver.cc]: check if common code can be extracted in a file in parent dir instead duplicate it, similar to collection.cc class

File: lib/tests/CMakeLists.txt [hunk 8]
```diff
@@ -75,7 +75,7 @@ if (UNIX)
         "util/posix/scoped_descriptor.cc"
     )
     if (OPENSSL_FOUND)
```
Comment (ciprianbadescu) [inline lib/tests/CMakeLists.txt]: same as src CMakeLists.txt

File: lib/inc/internal/facts/windows/ssh_resolver.hpp (hunk 5 above)
Comment (GeorgeMrejea) [reply lib/inc/internal/facts/windows/ssh_resolver.hpp]: I have took it under consideration, but refactoring to have single namespace might introduce some problem. I would avoid modify existing posix code

File: lib/CMakeLists.txt (hunk 6 above)
Comment (luchihoratiu) [reply lib/CMakeLists.txt]: Agreed and done.

File: lib/tests/CMakeLists.txt (hunk 8 above)
Comment (luchihoratiu) [reply lib/tests/CMakeLists.txt]: Done.

File: lib/inc/internal/facts/windows/ssh_resolver.hpp (hunk 5 above)
Comment (luchihoratiu) [reply lib/inc/internal/facts/windows/ssh_resolver.hpp]: Managed to do so, no problems seen so far, but I'm honestly not happy about the end result. This, combined with the extraction of common code in base class, made an ugly diff (also noisy due to changing paths/namespaces).
@ciprianbadescu : Please have a look at the 2nd commit and let me know if this is what you had in mind. Other suggestions/approaches are very welcomed! 😄

File: lib/src/facts/windows/ssh_resolver.cc (hunk 7 above)
Comment (luchihoratiu) [reply lib/src/facts/windows/ssh_resolver.cc]: Done.

**Pull request merged**